tbdex/rfq/rfq.go

```diff
@@ -0,0 +1,124 @@
+type RFQData struct {
+	OfferingID string               `json:"offeringId"`
+	Payin      SelectedPayinMethod  `json:"payin"`
+	Payout     SelectedPayoutMethod `json:"payout"`
+	ClaimsHash string               `json:"claimsHash,omitempty"`
+}
+
+// RFQPrivateData contains data which can be detached from the payload without disrupting integrity.
+type RFQPrivateData struct {
+	Salt   string                 `json:"salt"`
+	Claims []string               `json:"claims"`
+	Payin  *PrivatePaymentDetails `json:"payin"`
+	Payout *PrivatePaymentDetails `json:"payout"`
+}
+
+// PrivatePaymentDetails is a container for the cleartest [PaymentDetails]
+type PrivatePaymentDetails struct {
+	PaymentDetails map[string]any `json:"paymentDetails"`
+}
+
+// SelectedPayinMethod represents the chosen method for the pay-in
+type SelectedPayinMethod struct {
+	Amount             string `json:"amount"`
+	Kind               string `json:"kind"`
+	PaymentDetailsHash string `json:"paymentDetailsHash,omitempty"`
+}
+
+// SelectedPayoutMethod represents the chosen method for the pay-out
+type SelectedPayoutMethod struct {
+	Kind               string `json:"kind"`
+	PaymentDetailsHash string `json:"paymentDetailsHash,omitempty"`
+}
+
+// Digest computes a hash of the rfq
+func (r RFQ) Digest() ([]byte, error) {
+	payload := map[string]any{"metadata": r.MessageMetadata, "data": r.Data}
+
+	hashed, err := tbdex.DigestJSON(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to digest rfq: %w", err)
+	}
+
+	return hashed, nil
+}
+
+// Sign cryptographically signs the RFQ using DID's private key
+func (r *RFQ) Sign(bearerDID did.BearerDID) error {
+	r.MessageMetadata.From = bearerDID.URI
+
+	signature, err := tbdex.Sign(r, bearerDID)
+	if err != nil {
+		return fmt.Errorf("failed to sign rfq: %w", err)
+	}
+
+	r.Signature = signature
+
+	return nil
+}
+
+// UnmarshalJSON validates and unmarshals the input data into an RFQ.
+func (r *RFQ) UnmarshalJSON(data []byte) error {
+	err := tbdex.Validate(tbdex.TypeMessage, data, tbdex.WithKind(RFQKind))
+	if err != nil {
+		return fmt.Errorf("invalid rfq: %w", err)
+	}
+
+	ret := rfq{}
+	err = json.Unmarshal(data, &ret)
+	if err != nil {
+		return fmt.Errorf("failed to JSON unmarshal rfq: %w", err)
+	}
+
+	*r = RFQ(ret)
+
+	_, err = tbdex.VerifySignature(r, r.Signature)
+	if err != nil {
+		return fmt.Errorf("failed to verify RFQ signature: %w", err)
+	}
+
+	// TODO add check when decoded.SignerDID is implemented
+	// if decoded.SignerDID != r.MessageMetadata.From {
+	// 	return errors.New("signer: %w does not match message metadata from: %w", decoded.Header.SignerDID, r.MessageMetadata.From)
+	// }
+
+	// TODO verify private data
+	// if requirePrivateData {
+	// 	err = r.verifyAllPrivateData()
+	// 	if err != nil {
+	// 	return fmt.Errorf("failed to verify private data: %w", err)
+
+	// 	}
+	// }
+
+	return nil
+}
+
+func (r *RFQ) verifyAllPrivateData() error {
+
+	return nil
+}
+
+type rfq RFQ
```

Comment:
go n00b question: what's the point of this line? I see that we use `rfq` in `UnmarshalJSON` above, but why not use `RFQ`?

Reply:
good q @diehuxx . it's because of [this](https://github.com/TBD54566975/tbdex-go/blob/3f387f010b889df7cc3cba0e3369b803a8573f66/tbdex/rfq/rfq.go#L90) line in `Unmarshal`.  we use a net new type definition in order to prevent `Unmarshal` from infinitely looping. without the additional type definition, the same `Unmarshal` gets called over and over and over

Reply:
unmarshal all the way down